src/solvers/dgmulti/flux_differencing.jl

```diff
@@ -13,69 +13,99 @@
+@inline function hadamard_sum!(du, A,
                                flux_is_symmetric::Val{true}, volume_flux,
-                               orientation, u, equations)
+                               orientation_or_normal_direction, u, equations)
+  row_ids, col_ids = axes(A)
 
-  rows, cols = axes(A)
-  for i in rows
+  for i in row_ids
     u_i = u[i]
     du_i = du[i]
-    for j in cols
+    for j in col_ids
       # This routine computes only the upper-triangular part of the hadamard sum (A .* F).
       # We avoid computing the lower-triangular part, and instead accumulate those contributions
       # while computing the upper-triangular part (using the fact that A is skew-symmetric and F
       # is symmetric).
       if j > i
-          AF_ij = A[i,j] * volume_flux(u_i, u[j], orientation, equations)
-          du_i = du_i + AF_ij
-          du[j] = du[j] - AF_ij
+        u_j = u[j]
+        AF_ij = A[i,j] * volume_flux(u_i, u_j, orientation_or_normal_direction, equations)
+        du_i = du_i + AF_ij
+        du[j] = du[j] - AF_ij
       end
     end
     du[i] = du_i
   end
 end
 
-# Version for dense matrices and non-symmetric fluxes
-@inline function hadamard_sum!(du, sparsity_pattern::Nothing, A,
+# Version for dense operators and non-symmetric fluxes
+@inline function hadamard_sum!(du, A,
+                               flux_is_symmetric::Val{false}, volume_flux,
+                               orientation::Integer, u, equations)
+  row_ids, col_ids = axes(A)
+
+  for i in row_ids
+    u_i = u[i]
+    du_i = du[i]
+    for j in col_ids
+      u_j = u[j]
+      f_ij = volume_flux(u_i, u_j, orientation, equations)
+      du_i = du_i + A[i,j] * f_ij
+    end
+    du[i] = du_i
+  end
+end
+
+# TODO: DGMulti. Fix for curved meshes.
+@inline function hadamard_sum!(du, A,
                                flux_is_symmetric::Val{false}, volume_flux,
-                               orientation, u, equations)
-  rows, cols = axes(A)
-  for i in rows
+                               normal_direction::AbstractVector, u, equations)
+  row_ids, col_ids = axes(A)
+
+  for i in row_ids
     u_i = u[i]
     du_i = du[i]
-    for j in cols
-        f_ij = volume_flux(u_i, u[j], orientation, equations)
-        du_i = du_i + A[i,j] * f_ij
+    for j in col_ids
+      u_j = u[j]
+      # The `normal_direction::AbstractVector` has to be passed in twice.
+      # This is because on curved meshes, nonconservative fluxes are
+      # evaluated using both the normal and its average at interfaces.
+      f_ij = volume_flux(u_i, u_j, normal_direction, normal_direction, equations)
+      du_i = du_i + A[i,j] * f_ij
     end
     du[i] = du_i
   end
 end
 
-# Version for symmetric fluxes and sparse operators
-@inline function hadamard_sum!(du, sparsity_pattern::AbstractSparseMatrix{Bool}, A,
+# Version for sparse operators and symmetric fluxes
+@inline function hadamard_sum!(du, A::LinearAlgebra.Adjoint{<:Any, <:SparseMatrixCSC},
```

Comment:
Is there a reason why we don't use `A::LinearAlgebra.Adjoint{<:Any, <:AbstractSparseMatrix}` (or `A::LinearAlgebra.Adjoint{<:Any, <:AbstractSparseMatrixCSC}` if we're worried about performance)?

Reply:
Good point - fixed in a291133a7cac379b46ff1f04b1c0a182a738188a by switching to `AbstractSparseMatrixCSC`.